docs/AGENTS-ui.md

```diff
@@ -0,0 +1,524 @@
+
+  // Optional: Declare dependencies on other plugins
+  static readonly dependencies = [OtherPlugin];
+
+  // Called when the plugin is activated (before trace load)
+  static onActivate(app: App): void {
+    // Register commands, sidebar items, pages that don't need a trace
+  }
+
+  // Called when a trace is loaded
+  async onTraceLoad(trace: Trace): Promise<void> {
+    // Register tracks, tabs, commands that need trace data
+    // Query the trace processor, add tracks to the workspace
+  }
+}
+```
+
+**Plugin Lifecycle:**
+
+1. `onActivate()` - Called when the plugin is enabled, before any trace is loaded. Use for registering global commands, pages, and sidebar items.
+2. `onTraceLoad()` - Called when a trace is loaded. Use for registering tracks, tabs, and commands that depend on trace data.
+3. `trace.onTraceReady` event - Fired after all plugins have finished `onTraceLoad()`. Use for automations that need all tracks to be available.
+
+**Key APIs available to plugins:**
+
+- `trace.engine` - Run SQL queries against TraceProcessor
+- `trace.tracks` - Register and find tracks
+- `trace.selection` - Manage selection state
+- `trace.commands` - Register commands
+- `trace.tabs` - Register tabs in the details panel
+- `trace.timeline` - Access timeline state
+- `trace.workspace` - Manage the track tree structure
+
+## Mithril Patterns and Best Practices
+
+The UI uses Mithril.js. Follow these patterns:
+
+**Component Structure:**
+```typescript
+import m from 'mithril';
+
+interface MyComponentAttrs {
+  readonly value: string;
+  readonly onChange: (newValue: string) => void;
+}
+
+export class MyComponent implements m.ClassComponent<MyComponentAttrs> {
+  // Local state
+  private expanded = false;
+
+  view({attrs}: m.CVnode<MyComponentAttrs>): m.Children {
+    return m('.my-component',
+      m(Button, {label: attrs.value, onclick: () => this.expanded = !this.expanded}),
+      this.expanded && m('.details', 'Expanded content'),
+    );
+  }
+}
+```
+
+**Mithril Rules:**
+
+- No need to call`m.redraw()` from within lifecycle hooks or handlers. Mithril automatically schedules redraws.
+- Use `constructor` for initialization if no DOM access is needed, or `onCreate` if DOM is needed.
+- Prefer using the existing widget library (`ui/src/widgets/`) over creating new components.
+- Use `readonly` for attrs properties to prevent accidental mutation. We like things to be immutable.
+
+**Conditional Rendering with State Preservation:**
+Use the `Gate` component when you need to conditionally show/hide content while preserving component state:
+```typescript
+import {Gate} from '../base/mithril_utils';
+
+m(Gate, {open: this.isVisible}, m(ExpensiveComponent));
+```
+
+### Widget Library
+
+The `ui/src/widgets/` directory contains reusable components. Always check here before creating new UI elements:
+
+- `Button`, `ButtonBar`, `ButtonGroup` - Various button styles
+- `Menu`, `MenuItem`, `MenuDivider` - Dropdown menus
```

Comment:
Probably makes sense to mention PopupMenu here which is usually where these widgets are used